### PR TITLE
[4.0] Messages/Alerts: using icons instead of text as heading

### PR DESCRIPTION
**Pull request by infograf768**
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/30404

### Summary of Changes
As title says, replacing text by icons in Alerts

### Testing Instructions
Add to Atum and Cassiopea index.php

```
Factory::getApplication()->enqueueMessage('error message','error');
Factory::getApplication()->enqueueMessage('another long error message','error');
Factory::getApplication()->enqueueMessage('a danger message','danger');
Factory::getApplication()->enqueueMessage('warning message','warning');;
Factory::getApplication()->enqueueMessage('notice message','notice');
Factory::getApplication()->enqueueMessage('info message','info');
Factory::getApplication()->enqueueMessage('success message','success');
Factory::getApplication()->enqueueMessage('message no type');
```
in order to test all kinds of messages

Patch, Run NPM CI

### Actual result BEFORE applying this Pull Request

Screen Shot 2020-08-30 at 09 38 28

### Expected result AFTER applying this Pull Request
Cassiopea

Screen Shot 2020-08-30 at 09 23 41

Atum

Screen Shot 2020-08-30 at 09 23 22

### CODE REVIEW

**Comment by brianteeman** [comment]
Not got time to check properly but this looks ok for accessibility at first glance. Needs to be reviewed by the accessibility team though.

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on dce6e70204bf5fcd2f541b1ef878301c81436ac3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30516">issues.joomla.org/tracker/joomla-cms/30516</a>.</sub>

**Comment by brianteeman** [comment]
After applying the patch and rebuilding npm the pr itself works great but it has a knock on impact elsewhere

### before
image

### after
image

**Comment by infograf768** [comment]
> After applying the patch and rebuilding npm the pr itself works great but it has a knock on impact elsewhere

Yep, we saw that a bit ago.
Changing to draft.

**Comment by infograf768** [comment]
icomoon stuff was loaded, creating the issue with the featured columns icon.
It is now solved, but we have another issue:
The icons are very slow to load, including for all the other icons than the alerts ones.

Maybe this is due to the use of `@extend` in, for example
```
    .error,
    .danger {
      @extend .fas;
      @extend .fa-times-circle;
    }
```
and the fact that it is necessary if we use `@extend` to import fontawesome in joomla-alert.scss


We may have use a simpler
```
.error,
.danger {
      &::before {
        font-family: "Font Awesome 5 Free";
        font-weight: 900;
        content: "\f057";
      }
```
etc.

**Comment by ReLater** [comment]
Before this PR I had the possibility to use custom language strings for custom message types
`Factory::getApplication()->enqueueMessage('message my type', 'blah');`

After this PR not.

---------------
both template.scss are importing already fontawesome. Why twice with this pr?

https://github.com/joomla/joomla-cms/blob/4.0-dev/templates/cassiopeia/scss/template.scss#L15-L18

**Comment by infograf768** [comment]
@ReLater 
> both template.scss are importing already fontawesome. Why twice with this pr?

My test with `@extend` created errors when compiling as `@extend` can only work with selectors.
To solve this I had to load again fontawesome in the file where `@extend` is used.
This is why I need help from a scss expert. Will test without `@extend` to see how it behaves.

> Before this PR I had the possibility to use custom language strings for custom message types

I guess you were always getting an info type message (blue background), with `blah` translated in the heading. 

Here we use icons related to core types.
As `blah` is not a specific type, it will show without a specific icon, unless you define a css for an icon in your template css.
But the text will be correctly translated.

Example:
`Factory::getApplication()->enqueueMessage('message with blah as type', 'blah');`

`BLAH="My custom heading text"`

Screen Shot 2020-08-30 at 16 55 20

**Comment by N6REJ** [comment]
This brings up an important issue I have had with the way icons are currently handled.  Which me'n hans are working on.  Give me 24hrs to sort this out for the best solution.  One thing we don't need/want is to load fa set twice.

**Comment by dgrammatiko** [comment]
@infograf768 for the love of god please don't mix Fontawesome with the alerts. They are and supposed to be universal (eg no dependencies and work well with any front end stack). Just copy paste these few lines of CSS from this comment: https://github.com/joomla-projects/custom-elements/issues/99#issuecomment-423287233

**Comment by infograf768** [comment]
@dgrammatiko 
I understand it should not be done this way as anyway the icons loading time is a no no.
But I am not sure how to apply your suggestion here.

Note: let's not forget that we have 2 ways to trigger the alerts:
1. the layout
2. the js

Therefore we do need a common resulting css for both.

**Comment by infograf768** [comment]
Obvioulsy, it is possible to solve the issue for the layout, using a simple code similar to the one used in https://github.com/brianteeman/joomla-cms/commit/8f6c3f97b35de5245b35886814ebabdb9b01c94a
This because we do not use the `h4` anymore.

i.e. adding
```
$icon = [
		CMSApplication::MSG_EMERGENCY => 'fas fa-times-circle',
		CMSApplication::MSG_ALERT     => 'fas fa-times-circle',
		CMSApplication::MSG_CRITICAL  => 'fas fa-times-circle',
		CMSApplication::MSG_ERROR     => 'fas fa-times-circle',
		CMSApplication::MSG_WARNING   => 'fas fa-exclamation-triangle',
		CMSApplication::MSG_NOTICE    => 'fas fa-info-circle',
		CMSApplication::MSG_INFO      => 'fas fa-info-circle',
		CMSApplication::MSG_DEBUG     => 'fas fa-info-circle',
		'message'                     => 'fas fa-check-circle',
		'success'                     => 'fas fa-check-circle',
		'danger'                      => 'fas fa-times-circle',
];
```
and then
```
<span class="<?php echo $icon[$type]; ?>" aria-hidden="true"></span>
```

But the problem remains for the js as I guess it is not a good idea to hardcode there these classes.

**Comment by brianteeman** [comment]
That is not correct for the reasons stated in that pr when it was closed

**Comment by infograf768** [comment]
I think I understood Dimitris proposal. On it now.

**Comment by dgrammatiko** [comment]
> I think I understood Dimitris proposal. On it now.

The same approach SHOULD be applied also to all the fields. Fields should be coded in such a way that Bootstrap/jQuery or Font Awesome is not a dependency... (eg icons should be a css baground-image property if you want the CMS to remain relative in terms of performance)

**Comment by infograf768** [comment]
@dgrammatiko @brianteeman 
Modified to use svg. No more issue loading time.

We now get

Screen Shot 2020-08-31 at 13 36 35

**Comment by infograf768** [comment]
Have to correct a lot of scss cs, does not prevent from testing ;)

File: build/media_source/system/js/core.es6/message.es6 [hunk 1]
```diff
@@ -77,7 +77,7 @@
       if (typeof title !== 'undefined') {
         titleWrapper = document.createElement('div');
         titleWrapper.className = 'alert-heading';
-        titleWrapper.innerHTML = Joomla.Text._(type) ? Joomla.Text._(type) : type;
+        titleWrapper.innerHTML = `<span class="${type}" aria-hidden ="true"></span><span class="sr-only">${Joomla.Text._(type) ? Joomla.Text._(type) : type}</span>`;
```
Comment (dgrammatiko) [inline build/media_source/system/js/core.es6/message.es6]: You don't need `aria-hidden ="true"` anymore as the icons are background images and will never be announced from screen readers

**Review by dgrammatiko — CHANGES_REQUESTED** [review]

File: build/media_source/system/js/core.es6/message.es6 (hunk 1 above)
Comment (dgrammatiko) [reply build/media_source/system/js/core.es6/message.es6]: Also there is a typo there, the equal has space before which probably will be ignored by browsers

Comment (infograf768) [reply build/media_source/system/js/core.es6/message.es6]: took off aria-hidden from both the js and the layout

**Comment by N6REJ** [comment]
why is the background size so big?
```css
background-image:url('data:image/svg+xml;utf8,<svg width="1792" height="1792" viewBox="0 0 1792 1792" xmlns="http://www.w3.org/2000/svg"><path fill="rgba(255, 255, 255, .95)" d="M1299 813l-422 422q-19 19-45 19t-45-19l-294-294q-19-19-19-45t19-45l102-102q19-19 45-19t45 19l147 147 275-275q19-19 45-19t45 19l102 102q19 19 19 45t-19 45zm141 83q0-148-73-273t-198-198-273-73-273 73-198 198-73 273 73 273 198 198 273 73 273-73 198-198 73-273zm224 0q0 209-103 385.5t-279.5 279.5-385.5 103-385.5-103-279.5-279.5-103-385.5 103-385.5 279.5-279.5 385.5-103 385.5 103 279.5 279.5 103 385.5z"/></svg>');
```

**Comment by infograf768** [comment]
@Quy 
On it.

**Comment by infograf768** [comment]
@Quy
Should be OK now.

@N6REJ 
> why is the background size so big?

I picked 2 of these at https://github.com/joomla-projects/custom-elements/issues/99#issuecomment-423287233
Then  I got the others from fontawesome svg and a bit mimicked code.

File: layouts/joomla/system/message.php [hunk 2]
```diff
@@ -45,7 +45,10 @@
 			<?php foreach ($msgList as $type => $msgs) : ?>
 				<joomla-alert type="<?php echo $alert[$type] ?? $type; ?>" dismiss="true">
 					<?php if (!empty($msgs)) : ?>
-						<div class="alert-heading"><?php echo Text::_($type); ?></div>
+						<div class="alert-heading">
+							<span class="<?php echo $type; ?>"></span>
```
Comment (Quy) [inline layouts/joomla/system/message.php:49]:
```suggestion
							<span class="<?php echo $type; ?>" aria-hidden="true"></span>
```

File: build/media_source/system/js/core.es6/message.es6 [hunk 3]
```diff
@@ -77,7 +77,7 @@
       if (typeof title !== 'undefined') {
         titleWrapper = document.createElement('div');
         titleWrapper.className = 'alert-heading';
-        titleWrapper.innerHTML = Joomla.Text._(type) ? Joomla.Text._(type) : type;
+        titleWrapper.innerHTML = `<span class="${type}"></span><span class="sr-only">${Joomla.Text._(type) ? Joomla.Text._(type) : type}</span>`;
```
Comment (Quy) [inline build/media_source/system/js/core.es6/message.es6:80]:
```suggestion
        titleWrapper.innerHTML = `<span class="${type}" aria-hidden="true"></span><span class="sr-only">${Joomla.Text._(type) ? Joomla.Text._(type) : type}</span>`;
```

File: layouts/joomla/system/message.php (hunk 2 above)
Comment (dgrammatiko) [reply layouts/joomla/system/message.php:49]: @Quy aria-hidden is not needed here, the icon is a background image and thus will never be announced by a screen reader, https://github.com/joomla/joomla-cms/pull/30516#discussion_r480076023

Comment (Quy) [reply layouts/joomla/system/message.php:49]: Of course,..thanks!

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on c39992ef2fd865aae6e84f0f17dc380a93cba50f<br /><br />Tested on win10 with FF and Edged. Just a hint for other testers: my cache was stubborn for these changes, so using a new instance of browsers is agood idea.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30516">issues.joomla.org/tracker/joomla-cms/30516</a>.</sub>

**Comment by dgrammatiko** [comment]
I have tested this item :white_check_mark: successfully on c39992ef2fd865aae6e84f0f17dc380a93cba50f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30516">issues.joomla.org/tracker/joomla-cms/30516</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30516">issues.joomla.org/tracker/joomla-cms/30516</a>.</sub>

**Comment by HLeithner** [comment]
Thanks

**Pull request merged**